app/controllers/api/v0/versions_controller.rb

```diff
@@ -25,6 +27,41 @@ def show
     }
   end
 
+  def raw
+
+    @version ||= version_collection.find(params[:id])
+
+    def self.public_hosts=(hosts)
+      hosts = [] if hosts.nil?
+      hosts = hosts.split(' ') if hosts.is_a?(String)
+      unless hosts.is_a?(Enumerable) && hosts.all? {|host| host.is_a?(String)}
+        raise StandardError, 'Public hosts must be a string or enumerable of strings'
+      end
+      @public_hosts = hosts
+    end
+
+    if ENV['PUBLIC_ARCHIVE_HOSTS']
+      self.public_hosts = ENV['PUBLIC_ARCHIVE_HOSTS']
+    end
+
+    def self.public_hosts
+      @public_hosts || []
+    end
+
+    def self.public_archive_uri?(uri)
+      public_hosts.any? {|base| uri.starts_with?(base)}
+    end
+
+    if public_archive_uri?(@version.uri)
+      redirect_to @version.uri, status: 301 and return
+    else
+      upstream = open(@version.uri)
```

Comment:
See above about feature and security issues with `open-uri`.

If the `uri` is not public, I think the right thing here might actually be calling `Archiver.store.get_file()`.

Reply:
That seems right to me too. Done.